enhancements/implicit-catalog-versioning.md

```diff
@@ -0,0 +1,190 @@
+│   ├── etcdbackup.crd.yaml
+│   ├── etcdcluster.crd.yaml
+│   ├── etcdoperator.clusterserviceversion.yaml
+│   └── etcdrestore.crd.yaml
+└── metadata
+    └── annotations.yaml
+```
+
+In this format, it becomes significantly less clear how to define your update graph. If you are only ever publishing a specific version of your operator, how do you know how it fits into the graph? What has even been published previously? Where do you even get the metadata needed to define your update graph? The author of this operator for this trivial example would have to either piece together their previous releases based on a hopefully accurate source control history, or they would need to use some external tooling (which does not currently exist today) to inspect the already created Index image to see what has been released and define their graph that way.
+
+The primary motivation for this enhancement is to simplify all of that complexity and guesswork by redefining the main driver for the update graph to just use semantic versioning. If the versioning is implicit (i.e. 1.3.1 is an update from 1.3.0) all the operator author needs to be aware of is "is my new version the latest thing?" -- they do not need to inspect the manifests of existing versions to construct the graph.
+
+In the future, this proposal also helps to solve the ordering problem when building and modifying index images. Today when adding to Index images via the `opm index` command, you need to specify your images in the order that they replace previous images because of the possibility that the explicit data will make an invalid intermediate graph. With the use of semantic version, it becomes trivial to insert elements into the middle of the graph. It also allows a future in which specific versions can be deleted from graphs without a need to edit the yaml.
+
+### Goals
+
+* Define Catalog update graphs for individual operators based implicitly on semantic version rather than asking users to explicitly mention what version the previous version replaces
+* Update OLM to handle the new format of catalogs
+* Handle conversion of existing operator subscriptions to use the new format
+
+### Non-Goals
+
+* Modify any of the features of OLM or the Operator Registry to use new information from the semantic version. This proposal simply suggests using the semantic version to define the graph
+* Resolve the migration of existing subscriptions to use the new Catalog Index images
+* Make any explicit changes to existing OLM API versions (CSV will remain `v1alpha1`)
+
+## Proposal
+
+Basic approach:
+
+- Make the `spec.version` field required going forward when building registry images with `opm registry` commands
+- Modify the operator registry to build dependency graph based simply on `spec.version` field rather than using a combination of channel heads and `spec.replaces` for each graph
+- Modify OLM to use the `spec.version` field everywhere it attempts to parse the graph for update information
```

Comment:
Sorry, this is probably unclear. In the registry changes we did previously, we already made the change that each version explicitly subscribes to a set of one or more channels. So the query for "what set of versions can I update to" is actually more like "what set of versions can I update to given that I am subscribed to this channel"

I think that there is valuable information that we can gain in the future that starts with OLM explicitly learning this information (for example, we could create an option to update subscriptions to automatically update but only on patch version). It's probably fair to say that we could make this simpler right now in OLM by abstracting the graph questions into the registry APIs, but I would argue that down the line that we are just pushing that complexity into a project that eventually isn't chiefly concerned with it.

As an aside, I also tried to avoid any explicit tampering with the manifests in order to generate the update graphs that we want (and doing this feature this way allows us to make `skipRange` work in a more first class way that doesn't involve this sort of modification)